### PR TITLE
Polkadot governance api endpoints

### LINKED ISSUE #4445: polkadot governance api endpoints
Part of #4456 

- [x] governance overview
  - totalStakedAssets
  - totalVoters (all staked accounts)
  - treasurySize (treasure/communty pool size)
  - recentProposals (make adjustable pagesize, maybe extra endpoint)
  - topVoters (council members in Polkad and highst voting power in Cosmos)
  - links (title, link, type, icon)

- [x] ~add full text search to proposals endpoint~
- [x] ~make title, description, summary overwritable from db~
- [x] add polkadot proposals
  - add text proposals like in Cosmos
  - add treasury proposals via a templte for the description and summary
  - add parameter change proposals via a template
- [ ] add to proposals detailed votes
  - deposits
  - deposits sum
  - percentage of needed deposits
  - votes (add filter to only show for your validators)
  - votes sum
  - voting threshold yes
  - voting threshold no
  - links (from db, have title, link, icon, type)
  - timeline [{title, time}] (maybe put in network config)

### PR DESCRIPTION
**Pull request by mariopino**
Partially closes #4445 

**Description:**

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by mariopino** [comment]
Test query:

```gql
query proposals {
  proposals(networkId: "kusama") {
    id
    title
    type
    tally {
      yes
      no
      abstain
      veto
      total
      totalVotedPercentage
    }
    deposit
    status
    statusEndTime
  }
}
```

```json
{
  "data": {
    "proposals": [
      {
        "id": 73,
        "title": "0x5d0251bf0b0c1df02b2a2166719a1d0c1c964b9f1bfaaca07113a2cb2e3b603b",
        "type": "Simplemajority",
        "tally": {
          "yes": "8522.332900",
          "no": "110020.824000",
          "abstain": "0",
          "veto": "0",
          "total": "118543.156900",
          "totalVotedPercentage": 0.64
        },
        "deposit": "61993.500000",
        "status": "VotingPeriod",
        "statusEndTime": "Sat, 25 Jul 2020 14:06:10 GMT"
      }
    ]
  }
}
```

**Comment by mariopino** [comment]
```
const proposals = await api.derive.treasury.proposals()
console.log(JSON.stringify(proposals, null, 2))
```

```
{
  "approvals": [],
  "proposalCount": 39,
  "proposals": [
    {
      "council": [
        {
          "hash": "0x437283774530ccde4e9ca20ed3c8de486089c66c7c468f2830d6f8c7bddc5ad8",
          "proposal": {
            "callIndex": "0x1201",
            "args": {
              "proposal_id": 38
            }
          },
          "votes": {
            "index": 183,
            "threshold": 11,
            "ayes": [
              "GLVeryFRbg5hEKvQZcAnLvXZEXhiYaBjzSDwrXBXrfPF7wj",
              "J9nD3s7zssCX7bion1xctAF6xcVexcpy2uwy4jTm9JL8yuK",
              "Hjuii5eGVttxjAqQrPLVN3atxBDXPc4hNpXF6cPhbwzvtis",
              "EDkyLR1J19e9agASF6PK649jiheqD95tRweFNyzTg11Xug4",
              "H9eSvWe34vQDJAWckeTHWSqSChRat8bgKHG39GC1fjvEm7y",
              "DWUAQt9zcpnQt5dT48NwWbJuxQ78vKRK9PRkHDkGDn9TJ1j",
              "DTLcUu92NoQw4gg6VmNgXeYQiNywDhfYMQBPYg2Y1W6AkJF",
              "GcqKn3HHodwcFc3Pg3Evcbc43m7qJNMiMv744e5WMSS7TGn",
              "DfiSM1qqP11ECaekbA64L2ENcsWEpGk8df8wf1LAfV2sBd4"
            ],
            "nays": [],
            "end": 3312800
          }
        }
      ],
      "id": 38,
      "proposal": {
        "proposer": "DGfpcu129Bqp5PcWhharC4GusaXNi6SQv4PC1DUAE3Q6zWd",
        "value": 100000000000000,
        "beneficiary": "HR831wxo2N8jUPWVtJ2R8VviUjs34WAHeN6vp4biwHwiHpm",
        "bond": 5000000000000
      }
    }
  ]
}
```

**Comment by Bitcoinera** [comment]
Would love to get these amazing descriptions :sweat:  Still trying to figure it out looking at the polkassembly code:

https://kusama.polkassembly.io/motion/185

**Comment by faboweb** [comment]
🚀 Deployed on https://5f1ee5a4209ad7543acf602e--lunieio.netlify.app

File: api/lib/reducers/polkadotV0-reducers.js [hunk 1]
```diff
@@ -465,6 +465,158 @@ function rewardReducer(network, validators, reward, reducers) {
   return parsedRewards
 }
 
+function democracyProposalReducer(
+  network,
+  proposal,
+  totalIssuance,
+  blockHeight
+) {
+  return {
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:474]: ok for now, but we want to have a different structure in the future: https://github.com/luniehq/lunie/issues/4445 like `proposalsV2`

**Comment by Bitcoinera** [comment]
image

:eyes:

**Comment by Bitcoinera** [comment]
OK, I think it is only description now missing

**Comment by Bitcoinera** [comment]
Oops, Polkassembly broken:

image

File: api/lib/reducers/polkadotV0-reducers.js [hunk 2]
```diff
@@ -465,6 +465,159 @@ function rewardReducer(network, validators, reward, reducers) {
...
+    id: proposal.index,
+    networkId: network.id,
+    type: `text`,
+    title: `Democracy #${proposal.index}`,
```
Comment (jbibla) [inline api/lib/reducers/polkadotV0-reducers.js:478]:
```suggestion
    title: `Proposal #${proposal.index}`,
```

File: api/lib/reducers/polkadotV0-reducers.js [hunk 3]
```diff
@@ -465,6 +465,159 @@ function rewardReducer(network, validators, reward, reducers) {
...
+    description: proposal.description,
+    creationTime: undefined,
```
Comment (jbibla) [inline api/lib/reducers/polkadotV0-reducers.js:480]: not possible?

File: api/lib/reducers/polkadotV0-reducers.js [hunk 4]
```diff
@@ -465,6 +465,159 @@ function rewardReducer(network, validators, reward, reducers) {
...
+  network,
+  proposal,
+  totalIssuance,
+  blockHeight
+) {
+  return {
+    id: proposal.index,
+    network: network.id,
+    type: `text`,
+    title: `Referendum #${proposal.index}`,
```
Comment (jbibla) [inline api/lib/reducers/polkadotV0-reducers.js:498]:
```suggestion
    title: `Proposal #${proposal.index}`,
```

File: api/lib/reducers/polkadotV0-reducers.js [hunk 5]
```diff
@@ -465,6 +465,159 @@ function rewardReducer(network, validators, reward, reducers) {
...
+    proposer: undefined
+  }
+}
+
+function treasuryProposalReducer(network, proposal, councilMembers) {
+  return {
+    id: proposal.id,
+    networkId: network.id,
+    type: `text`,
+    title: `Treasury #${proposal.index}`,
```
Comment (jbibla) [inline api/lib/reducers/polkadotV0-reducers.js:515]:
```suggestion
    title: `Proposal #${proposal.index}`,
```

File: api/lib/reducers/polkadotV0-reducers.js [hunk 6]
```diff
@@ -465,6 +465,159 @@ function rewardReducer(network, validators, reward, reducers) {
...
+    deposit: toViewDenom(network, proposal.status.tally.turnout),
+    proposer: undefined
+  }
+}
+
+function treasuryProposalReducer(network, proposal, councilMembers) {
+  return {
+    id: proposal.id,
+    networkId: network.id,
+    type: `text`,
```
Comment (jbibla) [inline api/lib/reducers/polkadotV0-reducers.js:514]:
```suggestion
    type: `treasury`,
```

File: api/lib/reducers/polkadotV0-reducers.js [hunk 7]
```diff
@@ -465,6 +465,159 @@ function rewardReducer(network, validators, reward, reducers) {
...
+  network,
+  proposal,
+  councilMembers,
+  blockHeight
+) {
+  return {
+    id: proposal.votes.index,
+    networkId: network.id,
+    type: `text`,
+    title: `Council proposal #${proposal.votes.index}`,
```
Comment (jbibla) [inline api/lib/reducers/polkadotV0-reducers.js:534]:
```suggestion
    title: `Proposal #${proposal.votes.index}`,
```

File: api/lib/reducers/polkadotV0-reducers.js [hunk 8]
```diff
@@ -465,6 +465,159 @@ function rewardReducer(network, validators, reward, reducers) {
...
+function councilProposalReducer(
+  network,
+  proposal,
+  councilMembers,
+  blockHeight
+) {
+  return {
+    id: proposal.votes.index,
+    networkId: network.id,
+    type: `text`,
```
Comment (jbibla) [inline api/lib/reducers/polkadotV0-reducers.js:533]:
```suggestion
    type: `council`,
```

File: api/lib/reducers/polkadotV0-reducers.js (hunk 3 above)
Comment (faboweb) [reply api/lib/reducers/polkadotV0-reducers.js:482]: Maybe get the height and query the block for the time?

File: api/lib/reducers/polkadotV0-reducers.js (hunk 6 above)
Comment (faboweb) [reply api/lib/reducers/polkadotV0-reducers.js]: This should come from an enum

File: api/lib/reducers/polkadotV0-reducers.js (hunk 8 above)
Comment (faboweb) [reply api/lib/reducers/polkadotV0-reducers.js]: Should come from enum

Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js]: I'm on it :sweat_smile:

File: api/lib/reducers/polkadotV0-reducers.js (hunk 6 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js]: On it

File: api/lib/reducers/polkadotV0-reducers.js (hunk 3 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:482]: I need to check

File: api/lib/reducers/polkadotV0-reducers.js (hunk 7 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js]: You changed now your mind right?

File: api/lib/reducers/polkadotV0-reducers.js (hunk 3 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:496]: OK, this is basically what I have to do and after this I would call this ready

**Comment by Bitcoinera** [comment]
I'd call this ready depending on the criteria. Treasury proposals are for example missing all dates. Council proposals are missing creationTime. I can invest more time getting those, but it is not so WIP anymore

File: api/lib/reducers/polkadotV0-reducers.js [hunk 9]
```diff
@@ -5,6 +5,27 @@ const { lunieMessageTypes } = require('../../lib/message-types')
...
 const CHAIN_TO_VIEW_COMMISSION_CONVERSION_FACTOR = 1e-9
 
+const proposalTypeEnum = {
+  TEXT: 'TEXT',
+  COUNCIL: 'COUNCIL',
+  TREASURY: 'TREASURY',
+  PARAMETER_CHANGE: 'PARAMETER_CHANGE'
+}
+
+const proposalTypeEnumDictionary = (key) => {
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:15]: this looks weird. looks like individual parameters that get changed. is there a more general way to identify those proposals. or am I wrong?

File: api/lib/reducers/polkadotV0-reducers.js (hunk 1 above)
Comment (faboweb) [reply api/lib/reducers/polkadotV0-reducers.js:495]: we will do this in the old format and upgrade in a later PR

File: api/lib/reducers/polkadotV0-reducers.js [hunk 10]
```diff
@@ -470,6 +491,162 @@ function rewardReducer(network, validators, reward, reducers) {
...
+    total,
+    totalVotedPercentage: (total / councilMembers.length).toFixed(4) // the percent conversion is done in the FE. No need to multiply by 100
+  }
+}
+
+function democracyTallyReducer(proposal) {
+  // if we consider democracyProposals like the parallel to Cosmos proposals in deposit periods, then
+  // we would have to turn the seconds concept into a deposit somehow
+  return {
+    yes: proposal.seconds.length
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:634]: fine for now, we probably should display the strength of the proposals but feels very complex

File: api/lib/reducers/polkadotV0-reducers.js [hunk 11]
```diff
@@ -470,6 +491,162 @@ function rewardReducer(network, validators, reward, reducers) {
...
+  }
+}
+
+function toViewDenom(network, chainDenomAmount) {
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:638]: this is douplicated. can we refactor this function into a common lib or put it into the network object? (we need to have a network getter function because enriching the networks object currently needs to be done in a variety of places.)

File: api/lib/reducers/polkadotV0-reducers.js [hunk 12]
```diff
@@ -470,6 +491,162 @@ function rewardReducer(network, validators, reward, reducers) {
...
+  return BigNumber(chainDenomAmount)
+    .times(network.coinLookup[0].chainToViewConversionFactor)
+    .toFixed(6)
+}
+
+function getStatusEndTime(blockHeight, endBlock) {
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:644]:
```suggestion
// the status end time is a time "so and so days from the creation of the proposal opening"
function getStatusEndTime(blockHeight, endBlock) {
```

File: api/lib/source/cosmosV0-source.js [hunk 13]
```diff
@@ -238,14 +238,23 @@ class CosmosV0API extends RESTDataSource {
...
-      return this.reducers.proposalReducer(
-        this.network.id,
-        proposal,
-        {},
-        totalBondedTokens
-      )
-    })
+    const proposals = await Promise.all(
+      response.map(async (proposal) => {
+        const tally = await this.query(`/gov/proposals/${proposal.id}/tally`)
```
Comment (faboweb) [inline api/lib/source/cosmosV0-source.js:243]: do the queries in parallel

File: api/lib/source/polkadotV0-source.js [hunk 14]
```diff
@@ -529,6 +529,207 @@ class polkadotAPI {
       }
     )
   }
+
+  constructProposal = function (api, bytes) {
```
Comment (faboweb) [inline api/lib/source/polkadotV0-source.js:533]: weird syntax
```suggestion
  constructProposal(api, bytes) {
```

File: api/lib/source/polkadotV0-source.js [hunk 15]
```diff
@@ -529,6 +529,207 @@ class polkadotAPI {
...
+    const { data } = preimage.asAvailable
+    const proposalWithIndex = this.constructProposal(api, data)
+    const { meta, method } = api.registry.findMetaCall(
+      proposalWithIndex.callIndex
+    )
+    description = meta.documentation.toString()
+    proposalMethod = method
+
+    // get creationTime
+    const block = await api.rpc.chain.getBlock(blockHash)
```
Comment (faboweb) [inline api/lib/source/polkadotV0-source.js:569]: do as many requests in parallel that you can

File: app/src/components/governance/PageProposal.vue [hunk 16]
```diff
@@ -300,35 +309,13 @@ export default {
           )
         ) {
           this.found = true
+          this.loaded = true
+          return data.proposals.find(
```
Comment (faboweb) [inline app/src/components/governance/PageProposal.vue:313]: instead query for a single proposal from the API?

Comment (Bitcoinera) [reply app/src/components/governance/PageProposal.vue:313]: the problem is there is no way to query for single proposals in Polkadot apparently

Comment (Bitcoinera) [reply app/src/components/governance/PageProposal.vue:313]: and this query was here already, `proposals`, so it was kind of duplicating code

File: api/lib/source/polkadotV0-source.js (hunk 15 above)
Comment (Bitcoinera) [reply api/lib/source/polkadotV0-source.js:569]: Sadly it is not possible in the `getDemocracyProposalMetadata` and  `getReferendumProposalMetada`, as they chain all of the promise functions (they need the result to continue)

File: api/lib/source/polkadotV0-source.js (hunk 14 above)
Comment (Bitcoinera) [reply api/lib/source/polkadotV0-source.js:533]: yeah, coming straight from the polkadotjs repo :sweat_smile:

File: api/lib/reducers/polkadotV0-reducers.js (hunk 10 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:634]: this is rather the spirit of the detailedVotes PR

File: api/lib/reducers/polkadotV0-reducers.js (hunk 9 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:15]: I know, but I am afraid there is no other way :shrug: What I can do is to search in the repo for the list of all individual parameters that are changeable

File: api/lib/reducers/polkadotV0-reducers.js (hunk 11 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:638]: where do we need to enrich the network object concretely? You lost me there

Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:638]: I personally think it is fine to have `toViewDenom` as a common function in numbers.js

File: api/lib/reducers/polkadotV0-reducers.js (hunk 9 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:15]: Mmm, they seem to be dispersed in the code :shrug: these calls. Not so easy: https://crates.parity.io/pallet_staking/enum.Call.html#variant.set_validator_count

File: api/lib/reducers/polkadotV0-reducers.js (hunk 11 above)
Comment (faboweb) [reply api/lib/reducers/polkadotV0-reducers.js]: let's move it to numbers.js for now. I open an issue about the refactoring

Comment (faboweb) [reply api/lib/reducers/polkadotV0-reducers.js]: https://github.com/luniehq/lunie/issues/4664

File: app/src/components/governance/PageProposal.vue (hunk 16 above)
Comment (faboweb) [reply app/src/components/governance/PageProposal.vue:313]: we don't need to query for all proposals when just opening the proposal page right?

Comment (faboweb) [reply app/src/components/governance/PageProposal.vue:313]: so we can move the logic into the API even if we can not query for just one proposal in Polkadot.

Comment (faboweb) [reply app/src/components/governance/PageProposal.vue:313]: We should probably cash the proposals as well in store but in another PR

File: api/lib/reducers/polkadotV0-reducers.js (hunk 9 above)
Comment (faboweb) [reply api/lib/reducers/polkadotV0-reducers.js:18]: that doesn't look like a proposal but the resolution of a proposal that actually changes the parameter. you are sure this is the proposal?

File: app/src/components/governance/PageProposal.vue (hunk 16 above)
Comment (Bitcoinera) [reply app/src/components/governance/PageProposal.vue:313]: OK, gotcha

File: api/lib/source/polkadotV0-source.js [hunk 17]
```diff
@@ -725,7 +725,10 @@ class polkadotAPI {
     return orderBy(allProposals, 'id', 'desc')
   }
 
-  getProposalById() {}
+  async getProposalById(proposalId) {
```
Comment (faboweb) [inline api/lib/source/polkadotV0-source.js:728]: yep, nice

Comment (Bitcoinera) [reply api/lib/source/polkadotV0-source.js:728]: this is working now like it should. No need to query all proposals in the `PageProposal` component

File: api/lib/reducers/polkadotV0-reducers.js (hunk 9 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:18]: this is the proposal method. It is a metadata of the proposal itself that tells what method the proposal image will execute (new code patch that will be added to the runtime once the proposal is approved)

Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:18]: So what we could do is, to simplify, say if the proposal has a method then it is a Parameters Change proposal, but I am not 100% sure that the other ones have no method. Need to check. Maybe their method is something like "propose" or "plead" or something like that

Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:18]: Maybe all Polkadot proposals are actually parameters change :thinking:

Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js]: OK, I defaulted to PARAMETER_CHANGE since all democracy (and therefore referendas as well) start off with this "image" concept. So there is always going to be some new chunk of code getting approved

File: api/lib/reducers/polkadotV0-reducers.js [hunk 18]
```diff
@@ -482,7 +482,7 @@ function rewardReducer(network, validators, reward, reducers) {
 
 function democracyProposalReducer(network, proposal) {
   return {
-    id: Number(proposal.index),
+    id: `Democracy-`.concat(proposal.index),
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:485]:
```suggestion
    id: `democracy-`.concat(proposal.index),
```

File: api/lib/reducers/polkadotV0-reducers.js [hunk 19]
```diff
@@ -503,7 +503,7 @@ function democracyReferendumReducer(
   blockHeight
 ) {
   return {
-    id: Number(proposal.index),
+    id: `Referendum-`.concat(proposal.index),
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:506]:
```suggestion
    id: `referendum-`.concat(proposal.index),
```

File: api/lib/reducers/polkadotV0-reducers.js [hunk 20]
```diff
@@ -529,7 +529,7 @@ function treasuryProposalReducer(
   blockHeight
 ) {
   return {
-    id: Number(proposal.id),
+    id: `Treasury-`.concat(proposal.id),
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:532]:
```suggestion
    id: `treasury-`.concat(proposal.id),
```

File: api/lib/reducers/polkadotV0-reducers.js [hunk 21]
```diff
@@ -554,7 +554,7 @@ function councilProposalReducer(
   blockHeight
 ) {
   return {
-    id: Number(proposal.votes.index),
+    id: `Council-`.concat(proposal.votes.index),
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:557]:
```suggestion
    id: `council-`.concat(proposal.votes.index),
```

File: api/lib/reducers/polkadotV0-reducers.js (hunk 18 above)
Comment (faboweb) [reply api/lib/reducers/polkadotV0-reducers.js:485]: are there problems caching between networks? there could be 2 proposals on different networks with the same id

File: api/lib/reducers/polkadotV0-reducers.js [hunk 22]
```diff
@@ -1,10 +1,20 @@
...
-const { fixDecimalsAndRoundUp } = require('../../common/numbers.js')
+const {
+  fixDecimalsAndRoundUp,
+  toViewDenom
+} = require('../../common/numbers.js')
 const { lunieMessageTypes } = require('../../lib/message-types')
 
 const CHAIN_TO_VIEW_COMMISSION_CONVERSION_FACTOR = 1e-9
 
+const proposalTypeEnum = {
```
Comment (jbibla) [inline api/lib/reducers/polkadotV0-reducers.js:11]: in cosmosV2 this is called `proposalTypeEnumDictionary` ... can we make them the same?

File: api/lib/source/cosmosV0-source.js [hunk 23]
```diff
@@ -238,19 +238,28 @@ class CosmosV0API extends RESTDataSource {
...
-        {},
-        totalBondedTokens
-      )
-    })
+    const proposals = await Promise.all(
+      response.map(async (proposal) => {
+        const [tally, proposer] = await Promise.all([
+          this.query(`/gov/proposals/${proposal.id}/tally`),
+          this.query(`gov/proposals/${proposal.id}/proposer`).catch(() => {
+            return { proposer: `unknown` }
```
Comment (jbibla) [inline api/lib/source/cosmosV0-source.js:246]:
```suggestion
            return { }
```

Comment (jbibla) [reply api/lib/source/cosmosV0-source.js:246]: not sure about this but it seems odd to return `unknown`

File: api/lib/source/polkadotV0-source.js [hunk 24]
```diff
@@ -529,6 +529,213 @@ class polkadotAPI {
...
+    return {
+      ...proposal,
+      description,
+      proposer: proposal.proposer || proposer, // default to the already existing one if any
+      method: proposalMethod,
+      creationTime: proposal.creationTime || creationTime
+    }
+  }
+
+  async getReferendumProposalMetada(
```
Comment (jbibla) [inline api/lib/source/polkadotV0-source.js:585]:
```suggestion
  async getReferendumProposalMetadata(
```

File: api/lib/source/polkadotV0-source.js [hunk 25]
```diff
@@ -529,6 +529,213 @@ class polkadotAPI {
...
+      return await this.getDemocracyProposalMetadata(
+        proposal,
+        description,
+        proposer,
+        proposalMethod,
+        creationTime
+      )
+    }
+    if (type === `referendum`) {
+      return await this.getReferendumProposalMetada(
```
Comment (jbibla) [inline api/lib/source/polkadotV0-source.js:642]:
```suggestion
      return await this.getReferendumProposalMetadata(
```

File: app/src/components/governance/PageProposal.vue [hunk 26]
```diff
@@ -22,7 +22,9 @@
               Proposed by {{ proposal.validator.name }}:
               <Address :address="proposal.proposer" />
             </template>
-            <template v-else-if="proposal.proposer !== `unknown`">
+            <template
+              v-else-if="proposal.proposer && proposal.proposer !== `unknown`"
```
Comment (jbibla) [inline app/src/components/governance/PageProposal.vue:26]: if this was a blank object or null this would be easier

File: app/src/components/governance/LiProposal.vue [hunk 27]
```diff
@@ -15,7 +15,7 @@
       <h3 class="li-proposal-title">
         {{ proposal.title }}
       </h3>
-      <p class="li-proposal-description">
+      <p v-if="proposal.description" class="li-proposal-description">
```
Comment (jbibla) [inline app/src/components/governance/LiProposal.vue:18]: could there actually be no description?

Comment (Bitcoinera) [reply app/src/components/governance/LiProposal.vue:18]: yes, Treasury proposals don't have them. But true, I will just add a template, like the original idea was

File: app/src/components/governance/PageProposal.vue (hunk 26 above)
Comment (Bitcoinera) [reply app/src/components/governance/PageProposal.vue:26]: I can easily remove the `unknown` thing

File: api/lib/source/polkadotV0-source.js (hunk 25 above)
Comment (Bitcoinera) [reply api/lib/source/polkadotV0-source.js:642]: oops, thanks

File: api/lib/source/cosmosV0-source.js (hunk 23 above)
Comment (Bitcoinera) [reply api/lib/source/cosmosV0-source.js:246]: yeah, this was code from before actually. I also found it odd

File: api/lib/reducers/polkadotV0-reducers.js (hunk 22 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:11]: They are two different things, that is why they are called differently. In the cosmos source it is a real dictionary, converting from cosmos types to Lunie proposal types, while here it is just an enum

File: app/src/components/governance/PageProposal.vue (hunk 26 above)
Comment (Bitcoinera) [reply app/src/components/governance/PageProposal.vue:26]: Removed

File: app/src/components/governance/LiProposal.vue (hunk 27 above)
Comment (Bitcoinera) [reply app/src/components/governance/LiProposal.vue:18]: OK, now treasury also has a description coming from chain

File: api/lib/reducers/polkadotV0-reducers.js [hunk 28]
```diff
@@ -529,21 +529,22 @@ function treasuryProposalReducer(
   blockHeight
 ) {
   return {
-    id: `treasury-`.concat(proposal.id),
+    id: `treasury-`.concat(proposal.votes.index),
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:532]: `proposal.votes.index` is the unique identifier for the treasury proposal?

File: api/lib/source/polkadotV0-source.js [hunk 29]
```diff
@@ -702,14 +703,21 @@ class polkadotAPI {
         )
         .concat(
           treasuryProposals.proposals.map(async (proposal) => {
-            // should work. treasury and council proposals are virtually the same object
+            const treasuryProposal = proposal.council[0]
+              ? proposal.council[0]
+              : undefined
```
Comment (faboweb) [inline api/lib/source/polkadotV0-source.js:708]:
```suggestion
            const treasuryProposal = proposal.council[0]
```

File: api/lib/reducers/polkadotV0-reducers.js (hunk 28 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:532]: the same than in Council... there is another one if fact, but not used in Polkassembly. So I follow their advice as they are closely related to W3F

File: api/lib/source/polkadotV0-source.js (hunk 29 above)
Comment (Bitcoinera) [reply api/lib/source/polkadotV0-source.js:708]: true! It doesn't make sense. But why cannot I commit this suggestion?

Comment (jbibla) [reply api/lib/source/polkadotV0-source.js:708]: looks like a github bug 🐛

File: api/lib/source/polkadotV0-source.js [hunk 30]
```diff
@@ -702,14 +703,21 @@ class polkadotAPI {
         )
         .concat(
           treasuryProposals.proposals.map(async (proposal) => {
-            // should work. treasury and council proposals are virtually the same object
+            const treasuryProposal = proposal.council[0]
+              ? proposal.council[0]
```
Comment (jbibla) [inline api/lib/source/polkadotV0-source.js:707]:
```suggestion

```

File: api/lib/source/polkadotV0-source.js (hunk 29 above)
Comment (jbibla) [inline api/lib/source/polkadotV0-source.js:708]:
```suggestion

```

File: api/lib/reducers/polkadotV0-reducers.js (hunk 28 above)
Comment (jbibla) [reply api/lib/reducers/polkadotV0-reducers.js:532]: > there is another one if fact

another what?

File: app/src/components/governance/LiProposal.vue (hunk 27 above)
Comment (jbibla) [reply app/src/components/governance/LiProposal.vue:18]: from chain or from db?

File: app/src/components/governance/PageProposal.vue [hunk 31]
```diff
@@ -271,119 +279,83 @@ export default {
...
       update(data) {
-        /* istanbul ignore next */
-        return data.governanceParameters
+        return data.governanceParameters || {}
       },
+      /* istanbul ignore next */
       skip() {
-        /* istanbul ignore next */
-        return !this.found
+        // only Tendermint networks have this network-wide "governance parameters" logic
```
Comment (jbibla) [inline app/src/components/governance/PageProposal.vue:316]: wouldn't it be simpler to return an empty object or a message than to have network specific logic in the FE like this?

File: api/lib/reducers/polkadotV0-reducers.js (hunk 28 above)
Comment (faboweb) [reply api/lib/reducers/polkadotV0-reducers.js:532]: I don't understand what you wrote here, can you rephrase?

File: app/src/components/governance/PageProposal.vue (hunk 31 above)
Comment (faboweb) [reply app/src/components/governance/PageProposal.vue:316]: true, but we will always have protocol specific logic in the FE

Comment (jbibla) [reply app/src/components/governance/PageProposal.vue:316]: sure sure - but this feels like the wrong place to me. maybe up in the template...?

File: app/src/components/governance/PageProposals.vue [hunk 32]
```diff
@@ -23,6 +23,7 @@
       />
     </div>
     <ModalPropose
+      v-if="parameters && Object.keys(parameters).length > 0"
```
Comment (jbibla) [inline app/src/components/governance/PageProposals.vue:26]: what's this about?

File: api/lib/source/polkadotV0-source.js (hunk 29 above)
Comment (Bitcoinera) [reply api/lib/source/polkadotV0-source.js]: ? actually the change would be just leaving `proposal.council[0]` right?

File: app/src/components/governance/PageProposals.vue (hunk 32 above)
Comment (Bitcoinera) [reply app/src/components/governance/PageProposals.vue:26]: there are no governance parameters in Polkadot set, as it is always very flexible and varying according to the proposal

File: api/lib/source/polkadotV0-source.js (hunk 29 above)
Comment (Bitcoinera) [reply api/lib/source/polkadotV0-source.js]: ah ok, that is what you did. GitHub wasn't displaying very clearly

**Comment by Bitcoinera** [comment]
OK, this one is just really missing the real way to get description for referendum (the thing before was just a hack) and it will be done (for real).

Let's see if today we can merge this one and the other, #4607

File: app/src/components/governance/PageProposals.vue (hunk 32 above)
Comment (jbibla) [reply app/src/components/governance/PageProposals.vue:26]: sure but what does this have to do with hiding / showing the proposal modal?

Comment (Bitcoinera) [reply app/src/components/governance/PageProposals.vue:26]: the component is there, being called without v-if, which queries for stuff that Polkadot hasn't got.

We will really fix this once we implement the voting action, but for now this is better

File: api/lib/reducers/polkadotV0-reducers.js (hunk 28 above)
Comment (Bitcoinera) [reply api/lib/reducers/polkadotV0-reducers.js:533]: `proposal.votes.index is the unique identifier for the treasury proposal?` ---> this is the main one, yes

File: app/src/components/governance/LiProposal.vue (hunk 27 above)
Comment (Bitcoinera) [reply app/src/components/governance/LiProposal.vue:18]: Chain. We are not storing anything proposal-related yet, at least

**Comment by Bitcoinera** [comment]
Query to test (running API locally, of course):

```
query proposals {
  proposals(networkId: "kusama") {
    id
    title
    type
    tally {
      yes
      no
      abstain
      veto
      total
      totalVotedPercentage
    }
    description
    deposit
    status
    creationTime
    statusEndTime
    proposer
    validator {
      name
    }
  }
}
```

File: api/lib/source/cosmosV0-source.js [hunk 33]
```diff
@@ -238,19 +238,28 @@ class CosmosV0API extends RESTDataSource {
       '/staking/pool'
     )
     if (!Array.isArray(response)) return []
-    const proposals = response.map((proposal) => {
```
Comment (jbibla) [inline api/lib/source/cosmosV0-source.js:241]: why did this have to change?

File: api/lib/source/cosmosV0-source.js [hunk 34]
```diff
@@ -238,19 +238,28 @@ class CosmosV0API extends RESTDataSource {
...
-        this.network.id,
-        proposal,
-        {},
-        totalBondedTokens
-      )
-    })
+    const proposals = await Promise.all(
+      response.map(async (proposal) => {
+        const [tally, proposer] = await Promise.all([
+          this.query(`/gov/proposals/${proposal.id}/tally`),
```
Comment (jbibla) [inline api/lib/source/cosmosV0-source.js:244]: does this need a slash right at the beginning? line 245 does not have one...

File: app/src/components/governance/PageProposals.vue (hunk 32 above)
Comment (jbibla) [reply app/src/components/governance/PageProposals.vue:26]: can you leave a comment about future refactor then?

**Comment by jbibla** [comment]
- please respond to the remaining comments
- merge conflicts
- lint

- then we'll merge!

File: api/lib/source/cosmosV0-source.js (hunk 34 above)
Comment (Bitcoinera) [reply api/lib/source/cosmosV0-source.js:244]: ah good catch... this one was broken

File: api/lib/source/cosmosV0-source.js (hunk 33 above)
Comment (Bitcoinera) [reply api/lib/source/cosmosV0-source.js:241]: I don't know if I understand your question. It is better for performance to parallelize using Promise.all, so it is simply a performance improvement here

File: app/src/components/governance/PageProposal.vue (hunk 31 above)
Comment (Bitcoinera) [reply app/src/components/governance/PageProposal.vue:316]: template? What do you mean? I think this is a good place. Explains the skip

File: app/src/components/governance/PageProposals.vue (hunk 32 above)
Comment (Bitcoinera) [reply app/src/components/governance/PageProposals.vue:27]: Done

**Comment by Bitcoinera** [comment]
> * please respond to the remaining comments
> * merge conflicts
> * lint
> * then we'll merge!

Done. Let's merge!
